### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by KorenMary**
Added PatchNet with the integration tests

### CODE REVIEW

File: src/server/dcp_server/models.py [hunk 1]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+        self.classifier_train_config = classifier_train_config
+        self.classifier_eval_config = classifier_eval_config
+
+        # Initialize the classifier and the cellpose model 
+        self.classifier = CellFullyConvClassifier(
+            classifier_model_config, classifier_train_config,  classifier_eval_config)
+        self.segmentor = CustomCellposeModel(
+            segmentor_model_config, segmentor_train_config, segmentor_eval_config)
+
+    def init_from_checkpoints(self, chpt_classifier=None, chpt_segmentor=None):
```
Comment (christinab12) [inline src/server/dcp_server/models.py:226]: where is this function called?

File: src/server/dcp_server/models.py [hunk 2]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+                logits = self.classifier(img) 
+        
+                _, predicted = torch.max(logits, 1)
+                labels_fit.append(predicted)
+
+                # Assign predicted class to corresponding location in final_mask
+                final_mask[loc] = predicted + 1
+
+        # Apply mask to final_mask, retaining only regions where cellpose_mask is greater than 0
+        final_mask = final_mask * ((cellpose_mask > 0).long())
```
Comment (christinab12) [inline src/server/dcp_server/models.py:335]: why do we need this? won't locs give you coordinates of non-zero object locations?

File: src/server/dcp_server/models.py [hunk 3]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+        ## TODO: take care of the images and masks preparation -> this step isn't for now @Mariia
+        ## TODO call create_patches (adjust create_patch_dataset function)
+        ## to prepare imgs and masks for training CNN
+        ## TODO call self.classifier.train(imgs, masks)
+
+        num_classes = self.classifier_model_config.get("num_classes", 3)
+
+        black_bg = self.classifier_model_config.get("black_bg", False)
+        include_mask = self.classifier_model_config.get("include_mask", False)
+        include_cellpose_mask = self.classifier_model_config.get("include_cellpose_mask", True)
```
Comment (christinab12) [inline src/server/dcp_server/models.py:250]: I can't find where in the code these are used: black_bg, include_mask and include_cellpose_mask -> remove?

File: src/server/dcp_server/models.py [hunk 4]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+        )
+
+        self.layer3 = nn.Sequential(
+            nn.Conv2d(64, 128, 3, 2, 4),
+            nn.BatchNorm2d(128),
+            nn.ReLU(),
+            nn.Dropout2d(p=0.2),
+        )
+
+        self.final_conv = nn.Conv2d(128, self.num_classes, 1)
```
Comment (christinab12) [inline src/server/dcp_server/models.py:118]: should this be a fully connected layer rather than a convolution?

File: src/server/dcp_server/models.py [hunk 5]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+            imgs ():
+            masks ():
+            black_bg (bool): Flag indicating whether to use a black background for patches.
+            include_mask (bool): Flag indicating whether to include the mask along with patches.
+        '''
+
+        noise_intensity = kwargs.get("data", {}).get("noise_intensity", 5)
+
+        #max_patch_size = np.max([self.find_max_patch_size(mask) in masks])
+        max_patch_size = kwargs.get("data", {}).get("patch_size", 64)
```
Comment (christinab12) [inline src/server/dcp_server/models.py:520]: max_patch_size isn't called in this function. remove?

File: src/server/dcp_server/models.py [hunk 6]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+            # Check if the current patch size is larger than the maximum
+            if total_size > max_patch_size:
+                max_patch_size = total_size
+                max_patch_indices = obj
+            
+            max_patch_size_edge = np.ceil(np.sqrt(max_patch_size))
+
+            return max_patch_size_edge
+
+    def pad_centered_padded_patch(self, x: np.ndarray, c, p, mask: np.ndarray=None, noise_intensity=None) -> np.ndarray:
```
Comment (christinab12) [inline src/server/dcp_server/models.py:372]: I rename this function to crop_centered_padded_patch to avoid confusion

File: src/server/dcp_server/test_integration.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    segmentor_eval_config = read_config('eval', config_path='config.cfg')
+
+    patch_model = CellposePatchCNN(
+        segmentor_model_config, segmentor_train_config, segmentor_eval_config,
+        classifier_model_config, classifier_train_config, classifier_eval_config)
+    return patch_model
+
+@pytest.fixture
+def data_train():
+    images, masks = get_dataset("/home/ubuntu/data-centric-platform/src/server/dcp_server/data")
```
Comment (christinab12) [inline src/server/dcp_server/test_integration.py:61]: We cannot have hard coded paths in the tests, this should work across platforms. Replace this with running your code for creating a small sample of the synthetic dataset

File: src/server/dcp_server/test_integration.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        masks.append(msk)
+
+    return images, masks
+
+
+@pytest.fixture
+def patch_model():
+    classifier_model_config = {
+        "chpt_path":
+        "/home/ubuntu/data-centric-platform/src/server/dcp_server/data/classifier_checkpoint.pth"
```
Comment (christinab12) [inline src/server/dcp_server/test_integration.py:46]: We cannot have hard coded paths in the tests (only the configs can have hard coded paths), this should work across platforms. Why do we need to load a pertained model? You are training the model anyway in test_train_run right?

File: src/server/dcp_server/models.py
Comment (christinab12) [inline src/server/dcp_server/models.py:1]: @hpelin can you please take a look at CustomCellPose model? Any idea why a. in eval we call (and pass as arg) eval_config since we have self.eval_config in init and b. why we have deep copy of images in train?

**Comment by christinab12** [comment]
@KorenMary please take a look at my comments and changes made to the code and we can iterate once more on these

File: src/server/dcp_server/test.py
Comment (christinab12) [inline src/server/dcp_server/test.py:1]: not sure why this file is needed?

File: src/server/dcp_server/models.py (hunk 2 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py:316]: locs are the slices, hence final_mask[loc] = x, will always fill an entire region (rectangle given by slice) with the value of x. To fill out the empty regions multiply by cellpose mask.

File: src/server/dcp_server/models.py (hunk 3 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py]: We used it in the beginning, but I save it just in case we want to experiment. I remove it.

File: src/server/dcp_server/models.py [hunk 9]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+
+                # Assign predicted class to corresponding location in final_mask
+                final_mask[loc] = predicted + 1
+
+        # Apply mask to final_mask, retaining only regions where cellpose_mask is greater than 0
+        final_mask = final_mask * ((cellpose_mask > 0).long())
+        
+        return final_mask
+
+    def find_max_patch_size(self, mask):
```
[the remark this answers is not in the document]
Comment (KorenMary) [reply src/server/dcp_server/models.py:320]: I don't use it because in our dataset the maximum size is known, but it will be usefull for other datasets.

File: src/server/dcp_server/models.py (hunk 4 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py:112]: We use patches of the same size because we pad them, but we can also use it for patches of different sizes with batch_size=1 and lower learning rate. Additionally, I still think fully convolutional network is more flexible and has fewer parameters.

File: src/server/dcp_server/models.py (hunk 6 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py]: Thanks, it's better :)

File: src/server/dcp_server/models.py [hunk 10]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+
+        Args:
+            x (np.ndarray): The input array from which the patch will be cropped.
+            c (tuple): The coordinates (row, column, channel) at the center of the patch.
+            p (tuple): The size of the patch to be cropped (height, width).
+
+        Returns:
+            np.ndarray: The cropped patch with applied padding.
+        """
+        if mask.shape[0] < mask.shape[-1]:
```
[the remark this answers is not in the document]
Comment (KorenMary) [reply src/server/dcp_server/models.py:371]: Yes we need it to put the channel into the right position.

File: src/server/dcp_server/test_integration.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+def data_eval():
+    img = cv2.imread("/home/ubuntu/data-centric-platform/src/server/dcp_server/data/img.jpg", cv2.IMREAD_GRAYSCALE)
+    msk = np.load("/home/ubuntu/data-centric-platform/src/server/dcp_server/data/mask.npy")
+    return img, msk
+
+def test_train_run(data_train, patch_model):
+
+    images, masks = data_train
+    for _ in tqdm(range(1)):
+        loss_train = patch_model.train(deepcopy(images), deepcopy(masks))
```
[the remark this answers is not in the document]
Comment (KorenMary) [reply src/server/dcp_server/test_integration.py]: patch_model will change the images and masks while running, therefore pass the copy of those

File: src/server/dcp_server/test.py
Comment (KorenMary) [reply src/server/dcp_server/test.py:1]: Sorry, we don't need it, I used it to double-check and print things, I remove it.

File: src/server/dcp_server/models.py (hunk 1 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py:204]: I remove that

File: src/server/dcp_server/config.cfg [hunk 12]
```diff
@@ -1,26 +1,58 @@
...
+        },
+        "classifier":{
+            "in_channels": 3,
+            "num_classes": 3,
+            "black_bg": "False",
+            "include_mask": "False"
+        }
     },
+
     "data": {
```
Comment (christinab12) [inline src/server/dcp_server/config.cfg:28]: change to 'paths' and maybe move also model path here?

File: src/server/dcp_server/models.py [hunk 13]
```diff
@@ -60,12 +70,444 @@ def masks_to_outlines(self, mask):
...
         return utils.masks_to_outlines(mask) #[True, False] outputs
 
+class CellClassifierFCNN(nn.Module):
+    
+    '''
+    Fully convolutional classifier for cell images.
+
+    Args:
+        in_channels (int): Number of input channels.
+        num_classes (int): Number of output classes.
```
Comment (christinab12) [inline src/server/dcp_server/models.py:80]: update

File: src/server/dcp_server/models.py (hunk 1 above)
Comment (christinab12) [reply src/server/dcp_server/models.py:204]: thanks!

File: src/server/dcp_server/models.py (hunk 3 above)
Comment (christinab12) [reply src/server/dcp_server/models.py]: yes please remove whatever is not used, thanks!

File: src/server/dcp_server/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (christinab12) [reply src/server/dcp_server/models.py:320]: is there a place in the code where you would call this function and currently commented out?

File: src/server/dcp_server/models.py (hunk 4 above)
Comment (christinab12) [reply src/server/dcp_server/models.py:112]: okay thanks for the explanation

File: src/server/dcp_server/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (christinab12) [reply src/server/dcp_server/models.py:320]: here right? https://github.com/HelmholtzAI-Consultants-Munich/data-centric-platform/pull/30#discussion_r1362048795

File: src/server/dcp_server/models.py (hunk 5 above)
Comment (christinab12) [reply src/server/dcp_server/models.py]: okay I think it's best here to add an if statement, check if patch_size is given in the config and if not then call find_patch_size. Can you please implement this?

File: src/server/dcp_server/models.py [hunk 14]
```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
...
+            # m = (mask_ != central_label) & (mask_ > 0)
+            m = (mask_ != central_label) & (mask_ > 0)
+            x[m] = 0
+
+            if noise_intensity is not None:
+                x[m] = np.random.normal(scale=noise_intensity, size=x[m].shape)
+
+        patch = x[max(top, 0):min(bottom, x.shape[0]), max(left, 0):min(right, x.shape[1])]
+        
+        if len(c) == 3:
```
[the remark this answers is not in the document]
Comment (christinab12) [reply src/server/dcp_server/models.py:397]: we only call the crop_centered_padded_patch function in get_centered_patches, where c is passed as (cx, cy). Hence c can never have len=3. That's why I suggest to remove this if statement

File: src/server/dcp_server/models.py [hunk 15]
```diff
@@ -60,12 +70,444 @@ def masks_to_outlines(self, mask):
...
+        max_patch_size = self.train_config["classifier"]["train_data"]["patch_size"]
+        num_classes = self.train_config["classifier"]["train_data"]["num_classes"]
+
+        patches, labels = [], []
+        for img, mask_class, mask_instance in zip(imgs,  masks_classes, masks_instances):
+            # Convert to one-hot encoding
+            # mask has dimension WxHxNum_of_channels
+            patch, label = self.get_centered_patches(img, 
+                                                        mask_instance,
+                                                        self.train_config["classifier"]["train_data"]["patch_size"],
```
Comment (christinab12) [inline src/server/dcp_server/models.py:503]: change this to var max_patch size @KorenMary

File: src/server/dcp_server/models.py [hunk 16]
```diff
@@ -60,12 +70,444 @@ def masks_to_outlines(self, mask):
...
+            class_mask = class_mask * (instance_mask > 0)#.long())
+            final_mask = np.stack((instance_mask, class_mask)).astype(np.uint16)
+
+        self.eval_config['z_axis'] = 0
+
+        return final_mask
+
+    # REMOVE? replaced by code in eval
+    '''
+    def get_prediction(self, input_image, cellpose_mask):
```
Comment (christinab12) [inline src/server/dcp_server/models.py:272]: @KorenMary can you check if we can safely remove this function now and remove if so? Thanks!

File: src/server/dcp_server/models.py [hunk 17]
```diff
@@ -60,12 +70,444 @@ def masks_to_outlines(self, mask):
...
+        # train cellpose
+        masks = np.array(masks)
+        masks_instances = list(masks[:,0, ...])
+        self.segmentor.train(imgs, masks_instances)
+        # create patch dataset to train classifier
+        masks_classes = list(masks[:,1, ...])
+        patches, labels = self.create_patch_dataset(imgs, masks_classes, masks_instances)
+        # train classifier
+        self.classifier.train(patches, labels)
+        #return # TODO - define if we need to return something
```
Comment (christinab12) [inline src/server/dcp_server/models.py:236]: remove - nothing should be returned from train (see train of CustomCellposeModel)

File: src/server/dcp_server/models.py [hunk 18]
```diff
@@ -60,12 +70,444 @@ def masks_to_outlines(self, mask):
...
+        Args:
+            imgs ():
+            masks ():
+            black_bg (bool): Flag indicating whether to use a black background for patches.
+            include_mask (bool): Flag indicating whether to include the mask along with patches.
+        '''
+
+        noise_intensity = self.train_config["classifier"]["train_data"]["noise_intensity"]
+        max_patch_size = self.train_config["classifier"]["train_data"]["patch_size"]
+        num_classes = self.train_config["classifier"]["train_data"]["num_classes"]
```
Comment (christinab12) [inline src/server/dcp_server/models.py:495]: num_classes is not used here? remove?

File: src/server/dcp_server/config.cfg (hunk 12 above)
Comment (christinab12) [reply src/server/dcp_server/config.cfg:28]: @KorenMary please ignore this one, it is for me!

File: src/server/dcp_server/models.py (hunk 5 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py]: Done

File: src/server/dcp_server/models.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (KorenMary) [reply src/server/dcp_server/models.py:319]: "get_centered_patches" is used in both the train and eval functions.
In the train function, it's a part of the create_patch_dataset function, which is used during training. The code looks like this:
```
patches, labels = self.create_patch_dataset(imgs, masks_classes, masks_instances)
```
In this case, 'imgs' are 3-channel images, where the last dimension represents the number of channels.

In the eval function, the code is as follows:
```
patches, _ = self.get_centered_patches(img, instance_mask, self.eval_config["classifier"]["data"]["patch_size"], noise_intensity=5)
```
In this case, 'img' is an image with dimensions HxW.

File: src/server/dcp_server/test_integration.py (hunk 7 above)
Comment (KorenMary) [reply src/server/dcp_server/test_integration.py]: Changed it, now we generate data on-the-fly without saving anywhere.

File: src/server/dcp_server/test_integration.py (hunk 8 above)
Comment (KorenMary) [reply src/server/dcp_server/test_integration.py]: Yes, we train it from scratch, changed it.

File: src/server/dcp_server/models.py (hunk 13 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py:87]: Thanks :)

File: src/server/dcp_server/models.py (hunk 16 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py]: Done

File: src/server/dcp_server/models.py [hunk 19]
```diff
@@ -60,12 +70,444 @@ def masks_to_outlines(self, mask):
...
+            # get instance mask from segmentor
+            instance_mask = self.segmentor.eval(img)
+            # find coordinates of detected objects
+            locs = find_objects(instance_mask) 
+            class_mask = np.zeros(instance_mask.shape)
+            # get patches centered around detected objects
+            patches, _ = self.get_centered_patches(img, 
+                                                   instance_mask, 
+                                                   self.eval_config["classifier"]["data"]["patch_size"], 
+                                                   noise_intensity=5)
```
[the remark this answers is not in the document]
Comment (KorenMary) [reply src/server/dcp_server/models.py:259]: I agree. I added noise_intensity to eval config.

File: src/server/dcp_server/models.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (christinab12) [reply src/server/dcp_server/models.py:333]: get_centered patches is used in both train and eval functions yes. However `if len(c) == 3:` is written in `crop_centered_padded_patch` which is only called once like this:
```
patch = self.crop_centered_padded_patch(img.copy(),
                                       (c_x, c_y),
                                       (p_size, p_size),
                                       mask=mask,
                                       noise_intensity=noise_intensity)
```
this is why c can never have len=3, since c is only once given as (cx,cy). I'm removing this for now to avoid confusion

**Review by christinab12 — APPROVED** [review]
